### DISCUSSION TITLE
scan delay and rectangle box

### DISCUSSION THREAD

**Discussion by rubiyz** [mebjas/html5-qrcode#250] [Ideas]
Hello,

First thank you for this wonderfull library. Greatly appreciated.

I needed two distinct small features which I implemented on the base of the current version.

One is the possibility to insert a delay between successive scans. My customers need to scan many barcodes one after the other, but I haven't found a way to disable the scan for two seconds, the time needed to position the smartphone on the next product. 

The second is the possibility to specify a width and a height for the qrbox. My need is to scan 1d barcodes and it look better to have a rectangle to select 1d barcode rather than a square. 

So I implemented these two small features in html5-qrcode.ts and I was wondering if you might be interested to merge these into your library ?

Best regards,
Yves

**Comment by mebjas** [comment]
> One is the possibility to insert a delay between successive scans. My customers need to scan many barcodes one after the other, but I haven't found a way to disable the scan for two seconds, the time needed to position the smartphone on the next product.

What exactly is expected from the library for this case?

### Update (09/30/2021)

Support for rectangular scan area is now implemented in latest version of library - [Changelog](https://github.com/mebjas/html5-qrcode/blob/master/changelog.md#version-2013)

**Reply by kuhlaid** [reply]
Most stand-alone barcode scanners have a delay after they acquire a read on a code so it doesn't accidentally pick up something you do not want scanned as you move the scanner (or the object you are scanning). For example if you had a piece of paper filled with sets of barcodes, you want to scan ONE and then reposition the scanner to pickup another. However you DO NOT want to pickup other codes as you are moving the scanner across or off the page. Thus the delay. Ideally you could set the scanner to 'continuous read with delay' or 'read once until you press a button or something to start scanning again'.

**Reply by mebjas** [reply]
I see, I think this part can be better handled on the consumer side. You can always explicitly decide to not consume the other code.

**Comment by ROBERT-MCDOWELL** [comment]
@kuhlaid
this is your own coding to do what you wish, for example you can use a boolean to false until the scan is successful so you set on true until you finished to manage your data and go back to false for a new scan.

**Reply by rubiyz** [reply]
> 
> 
> @kuhlaid
> this is your own coding to do what you wish, for example you can use a boolean to false until the scan is successful so you set on true until you finished to manage your data and go back to false for a new scan.

Absolutely. It's a possibility to do it. I was just wondering if it might be a useful feature for others and if so, if it could be included directly in the library.

**Reply by ROBERT-MCDOWELL** [reply]
You can open a new isseu tagged as "new feature", maybe something like have a new setting called "scanDelay" to allow the user to control the new scan interval?

**Comment by Braggo** [comment]
> The second is the possibility to specify a width and a height for the qrbox. My need is to scan 1d barcodes and it look better to have a rectangle to select 1d barcode rather than a square.
> 
> So I implemented these two small features in html5-qrcode.ts and I was wondering if you might be interested to merge these into your library ?
> 
> Best regards,
> Yves

can u share ure implemented code for rectangle box?

thx :)

**Reply by benaloui** [reply]
@rubiyz can you please share with us you new implemented function (rezise rectangle for 1D barcodes) ? I'm trying to do that but I failed.
This is my e-mail if you wan to write me something : benaloui@gmail.com
Thank a lot dude.

**Comment by benaloui** [comment]
> 
> 
> Hello,
> 
> First thank you for this wonderfull library. Greatly appreciated.
> 
> I needed two distinct small features which I implemented on the base of the current version.
> 
> One is the possibility to insert a delay between successive scans. My customers need to scan many barcodes one after the other, but I haven't found a way to disable the scan for two seconds, the time needed to position the smartphone on the next product.
> 
> The second is the possibility to specify a width and a height for the qrbox. My need is to scan 1d barcodes and it look better to have a rectangle to select 1d barcode rather than a square.
> 
> So I implemented these two small features in html5-qrcode.ts and I was wondering if you might be interested to merge these into your library ?
> 
> Best regards,
> Yves

Hi @rubiyz, can you please share with us yours update code ?
BR

**Comment by rubiyz** [comment]
Hello,

I'm not too familiar with the best way to share the code. Anyway, I put the two files I modified here: https://github.com/rubiyz/html5-qrcode. Let me know if it helps or how I can do it otherwise.
Best regards,
Yves

**Comment by benaloui** [comment]
Thank you for your rapid answer.
can you just remember me how to compile again these update file to generate
a new html5-qrcode.min.

King regards.

**Reply by rubiyz** [reply]
basically you can convert the .ts to .js with the tsc command (https://www.typescriptlang.org/docs/handbook/compiler-options.html). Then I'm using uglifyjs to convert from .js to .min.js (https://www.npmjs.com/package/uglify-js). But again, I'm a beginner with javascript and typescript. There are probably other ways...

**Comment by benaloui** [comment]
Thank you very much. I'll try your method.
I'm also beginner with typescript.

**Comment by benaloui** [comment]
When I run my code after compilation a have this error :
[image: image.png]
Do you see an error in my source code here attached.

Can you send my your html5-qrcode.min.js to compare it with mine generated ?

Thanks

**Reply by rubiyz** [reply]
I was not able to see the image in your message. Have a look on the javascript files here: https://github.com/rubiyz/html5-qrcode/tree/main/output

**Reply by benaloui** [reply]
Hi @rubiyz do you have an example of script that works fine with yours update html5-qrcode.min.js ? Perhpas this could help me.

**Comment by benaloui** [comment]
Can you send me a screen shot of your result rectangle scanner ? Do you have an example of you code in with your update works ?

Thank you.

Envoyé de mon iPhone

**Reply by rubiyz** [reply]
basically it's a replacement for the original version and it should work without issue. To use the added fonctionnalities, the config should be changed to something like 

```
const config = {
      fps: 10,
      qrboxWidth: 240,
      qrboxHeight: 160,
      aspectRatio: 0.75,
      idleTimeout: 1500,
    };
```

**Reply by benaloui** [reply]
I'm using this constructor : 
var html5qrcode = new Html5Qrcode("qr-reader", false);

Do you use the same or this : 

var html5QrcodeScanner = new Html5QrcodeScanner(
	"reader", { fps: 10, qrbox: 250 });

Because I think it's not the same thing.

**Comment by rubiyz** [comment]
Here is my code, but there are other stuff that you would not need for 
sure. This code is meant to be used with Phoenix LiveView, but you 
should get the idea...
The pushEvent can be completely discarded for example. They are linked 
to LiveView.


import { Html5Qrcode } from "./html5-qrcode/html5-qrcode.js";
export const html5QRcode = {
   mounted() {
     var lastBarcode = "";

     this.handleEvent("stop", () => alert("stop"));

     var cameraId = null;

     // This method will trigger user permissions
     Html5Qrcode.getCameras()
       .then((devices) => {
         /**
          * devices would be an array of objects of type:
          * { id: "id", label: "label" }
          */
         if (devices && devices.length) {
           cameraId = devices[0].id;
           var element = 
document.getElementById("no.camera.placeholder");
           element.classList.add("hidden");

           this.pushEvent("camera_found", { devices });
           html5QrCode.start(
             { facingMode: { exact: "environment" } },
             config,
             qrCodeSuccessCallback
           );
         }
       })
       .catch((err) => {
         console.log("No camera found");
       });

     const html5QrCode = new Html5Qrcode("reader");
     const qrCodeSuccessCallback = (barcode) => {
       if (barcode != lastBarcode) {
         lastBarcode = barcode;
         setTimeout(function () {
           lastBarcode = "";
         }, 5000);
         this.pushEvent("barcode_scanned", { barcode });
       }
     };
     const config = {
       fps: 10,
       qrboxWidth: 240,
       qrboxHeight: 160,
       aspectRatio: 0.75,
       idleTimeout: 1500,
     };

     this.pushEvent("camera_selected", { cameraId });
   },
};


------ Original Message ------

**Comment by ROBERT-MCDOWELL** [comment]
why not to create a boolean with true when barcode is scanned until the next scann show a different barcode?

**Reply by rubiyz** [reply]
I guess that should do the job too. One edge case I can see is when the user want to scan twice in a row the same barcode, like in a shop to add the same product twice. The way I went with the delay doesn't have this issue, but might have other disadventages...

**Comment by mebjas** [comment]
PR for rectangular QR box - https://github.com/mebjas/html5-qrcode/pull/300

**Reply by mebjas** [reply]
Published with `version 2.0.13` - https://github.com/mebjas/html5-qrcode/releases/tag/v2.0.13

(on Github as well as npm)
